### PR TITLE
feat(node): carry out chunk_proof chunk against peer

### PR DESCRIPTION
**Pull request by maqi**
## Description

reviewpad:summary

### CODE REVIEW

File: sn_networking/src/event.rs [hunk 1]
```diff
@@ -1086,12 +1101,34 @@ impl SwarmDriver {
...
+            let all_keys = self
+                .swarm
+                .behaviour_mut()
+                .kademlia
+                .store_mut()
+                .record_addresses_ref();
+            let keys_to_fetch = self
+                .replication_fetcher
+                .add_keys(holder, keys_to_store, all_keys);
+            if !keys_to_fetch.is_empty() {
```
Comment (JasonPaulGithub) [inline sn_networking/src/event.rs]: write conditionals to a positive logic for better readability:

if keys_to_fetch.is_empty() 
{ trace!("no waiting keys to fetch from the network"); } 
else 
{ self.send_event(NetworkEvent::KeysToFetchForReplication(keys_to_fetch)); }

File: sn_node/src/node.rs [hunk 2]
```diff
@@ -628,3 +661,59 @@ impl Node {
...
+                key: key.clone(),
+                nonce,
+            });
+            let responses = network
+                .send_and_get_responses(&[peer_id], &request, true)
+                .await;
+            let n_verified = responses
+                .into_iter()
+                .filter_map(|(peer, resp)| {
+                    if let Ok(Response::Query(QueryResponse::GetChunkExistenceProof(Ok(proof)))) =
```
Comment (JasonPaulGithub) [inline sn_node/src/node.rs]: this is not critical but be careful of statements like this turning into what's known as an "arrow head" pattern.
[This](https://medium.com/@mena.meseha/how-to-refactor-the-arrow-type-code-8c40c21db85f) article can share more on this but in short, the best way to solve this is to extract the block of if statements into their own functions

**Review by JasonPaulGithub — APPROVED** [review]

**Comment by JasonPaulGithub** [comment]
comments not worth wasting time over, i have ODD tomorrow. back on Monday. Save time.
Please let me know you checked the comments @maqi

File: sn_networking/src/event.rs [hunk 3]
```diff
@@ -1217,6 +1252,72 @@ impl SwarmDriver {
             }
         }
     }
+
+    /// Check among all chunk type records that we have, select those close to the peer,
+    /// and randomly pick one as the verification candidate.
+    #[allow(clippy::mutable_key_type)]
+    fn select_verify_candiates(&mut self, peer: NetworkAddress) -> Vec<NetworkAddress> {
```
Comment (joshuef) [inline sn_networking/src/event.rs]: `select_verify_candiates` this naming is quite vague.
`select_verification_data_candidates` perhaps?

File: sn_node/src/node.rs [hunk 4]
```diff
@@ -410,6 +410,39 @@ impl Node {
...
+                    // Only report the node as bad when ALL the verification attempts failed.
+                    let mut attempts = 0;
+                    while attempts < 3 {
+                        if chunk_proof_verify_peer(&network, peer_id, &keys_to_verify).await {
+                            return;
+                        }
+                        // Replication interval is 22s - 45s.
+                        // Three times 15 gives 45s, which ensures at least one replication
+                        // shall be carried out during the period.
+                        tokio::time::sleep(std::time::Duration::from_secs(15)).await;
```
Comment (joshuef) [inline sn_node/src/node.rs]: can we clean this up and get the time in a documented const, please 🙇 (same with `attempts`)

**Review by joshuef — DISMISSED** [review]
Some small things to help with clarity, overall looking good though!

**Review by JasonPaulGithub — APPROVED** [review]

**Comment by JasonPaulGithub** [comment]
Hi @joshuef 
I see the requested changes have been resolved but the review is still marked as requested, so i'll dismiss the pending review and get this through. Hope that's ok

**Comment by JasonPaulGithub** [comment]
I'll also merge this in as nothing more is pending. Thanks chaps 👍

**Pull request merged**